### PR TITLE
Added extra use-case hints to the already present invariant definitions.

### PR DESCRIPTION
**Pull request by NlightNFotis**
This is just a dummy PR, against my own fork, so that people can review early the changes in the invariant code and documentation before they get submitted to develop.

### CODE REVIEW

File: src/util/invariant.h [hunk 1]
```diff
@@ -227,13 +227,23 @@ inline void invariant_violated_string(
 // The condition should only contain (unmodified) arguments to the method.
 // "The design of the system means that the arguments to this method
 //  will always meet this condition".
+//
+// The condition asserts properties that the function expects, including,
+// member variables of the object, in case they are implicit input to
+// the function. It is also supposed to signal that even though an object
+// is in a valid state, a particular function may not be operating on
+// objects with the particular properties this one enjoys.
```
Comment (hannes-steffenhagen-diffblue) [inline src/util/invariant.h]: enjoys?

File: src/util/invariant.h [hunk 2]
```diff
@@ -242,6 +252,10 @@ inline void invariant_violated_string(
 // changed by a previous method call.
 // "The contract of the previous method call means the following
 //  condition holds".
+//
+// A difference between CHECK_RETURN and POSTCONDITION is that CHECK_RETURN is
+// supposed to assert much more specific details about what a function has
+// returned, that is dependent on the specific needs of a caller.
```
Comment (hannes-steffenhagen-diffblue) [inline src/util/invariant.h]: The difference between `POST_CONDITION` and `CHECK_RETURN` isn't the details, it's that one is a statement of a function about itself, whereas the other is a statement about what a caller expects from a function.

File: src/util/invariant.h [hunk 3]
```diff
@@ -254,14 +268,21 @@ inline void invariant_violated_string(
 // This condition should be used to document that assumptions that are
 // made on goto_functions, goto_programs, exprts, etc. being well formed.
 // "The data structure is corrupt or malformed"
+//
+// It is okay for a DATA_INVARIANT to have an empty string as a reason string
+// if you can't state the reason, or doing so doesn't add any information -
+// just restates the condition.
```
Comment (hannes-steffenhagen-diffblue) [inline src/util/invariant.h:280]: IMHO doesn't belong as part of the documentation, also note that there are almost always better options.

File: src/util/invariant.h [hunk 4]
```diff
@@ -254,14 +268,21 @@ inline void invariant_violated_string(
...
   INVARIANT_STRUCTURED(CONDITION, TYPENAME, __VA_ARGS__)
 
 // Legacy annotations
 
 // The following should not be used in new code and are only intended
-// to migrate documentation and "error handling" in older code
+// to migrate documentation and "error handling" in older code.
+//
+// It is okay to use these annotations during development locally, but
+// make sure they are not present in any PR you are submitting.
```
Comment (hannes-steffenhagen-diffblue) [inline src/util/invariant.h]: That's a code/development style consideration that doesn't really belong here; In actual reality this is no different from "should not be used", obviously any code that's not made visible to other people doesn't fall under the style guidelines.

**Review by hannes-steffenhagen-diffblue — COMMENTED** [review]
I'm not really sure these changes are necessary

File: src/util/invariant.h (hunk 1 above)
Comment (NlightNFotis) [reply src/util/invariant.h]: yes, this is why I wanted a review. Maybe an Englishman can chime in? I don't know what a proper expression for that would be, but I don't think that what's there is unreasonable.

Comment (danpoe) [reply src/util/invariant.h]: Instead of adding the first part of this description, can we rephrase the existing one above? Maybe something like `The condition should only contain (unmodified) inputs to the method. Inputs include arguments passed to the method, as well as member variables that the
method may read.`

File: src/util/invariant.h [hunk 5]
```diff
@@ -227,13 +227,23 @@ inline void invariant_violated_string(
...
 #define PRECONDITION_STRUCTURED(CONDITION, TYPENAME, ...) \
   INVARIANT_STRUCTURED(CONDITION, TYPENAME, __VA_ARGS__)
 
 // The condition should only contain variables that will be returned /
 // output without further modification.
 // "The implementation of this method means that the condition will hold".
+//
+// A POSTCONDITION documents what the function can guarantee after a "normal"
+// execution about the state it returns to its caller (properties of objects
+// that it returns).
```
Comment (danpoe) [inline src/util/invariant.h]: Can we rephrase this to something like: `A POSTCONDITION documents what the method guarantees about its output when it returns, given that it was called with valid inputs. Outputs include the return value of the method, as well as member variables that the method may write.`

File: src/util/invariant.cpp [hunk 6]
```diff
@@ -116,21 +116,14 @@ void report_exception_to_stderr(const invariant_failedt &reason)
...
   out << "Invariant check failed\n"
-      << "File " << file
-      << " function " << function
-      << " line " << line << '\n'
+      << "File " << file << " function " << function << " line " << line << '\n'
       << "Condition: " << condition << '\n'
-      << "Reason: " << reason
-      << "\nBacktrace:\n"
+      << "Reason: " << reason << '\n'
+      << "Backtrace:" << '\n'
```
Comment (danpoe) [inline src/util/invariant.cpp]: Could include newline in string: `"Backtrace:\n"`

File: src/util/invariant.h [hunk 7]
```diff
@@ -160,10 +162,17 @@ invariant_violated_structured(
...
   Params &&... params)
 {
   std::string backtrace=get_backtrace();
-  ET to_throw(file, function, line, backtrace, std::forward<Params>(params)...);
+  ET to_throw(
+    file,
+    function,
+    line,
+    backtrace,
+    std::forward<Params>(params)...,
```
Comment (danpoe) [inline src/util/invariant.h]: Can we have the variable number of arguments as the last argument? (i.e., change the order of this line and the subsequent one). We also need to change the order of the `reason` and the `condition` parameters of the constructor of `invariant_failedt` then.

File: src/util/invariant.cpp (hunk 6 above)
Comment (NlightNFotis) [reply src/util/invariant.cpp]: Ahh no. It was like that but I changed it to make it more uniform and consistent with the rest of the stringbuilding process.

File: src/util/invariant.h [hunk 8]
```diff
@@ -225,15 +230,26 @@ inline void invariant_violated_string(
 // for INVARIANT.
 
 // The condition should only contain (unmodified) arguments to the method.
+// Inputs includ arguments passed to the function, as well as member
```
Comment (krobelus) [inline src/util/invariant.h]: typo, should be "include"

File: src/util/invariant.h [hunk 9]
```diff
@@ -242,6 +258,10 @@ inline void invariant_violated_string(
 // changed by a previous method call.
 // "The contract of the previous method call means the following
 //  condition holds".
+//
+// A difference between CHECK_RETURN and POSTCONDITION is that CHECK_RETURN is
+// a statement about the what the caller expects from a function, whereas a
```
Comment (krobelus) [inline src/util/invariant.h]: typo, "the" should be removed

**Review by danpoe — APPROVED** [review]
LGTM

File: src/util/invariant.cpp [hunk 10]
```diff
@@ -116,20 +116,14 @@ void report_exception_to_stderr(const invariant_failedt &reason)
...
+std::string invariant_failedt::what() const noexcept
 {
   std::ostringstream out;
   out << "Invariant check failed\n"
-      << "File " << file
-      << " function " << function
-      << " line " << line << '\n'
-      << "Reason: " << reason
-      << "\nBacktrace:\n"
+      << "File " << file << " function " << function << " line " << line << '\n'
```
Comment (krobelus) [inline src/util/invariant.cpp]: Since you're already touching this, maybe it's clearer to insert a comma after the function name, or use something like `<file:line>`

File: src/util/invariant.cpp [hunk 11]
```diff
@@ -116,20 +116,14 @@ void report_exception_to_stderr(const invariant_failedt &reason)
...
+      << "Condition: " << condition << '\n'
+      << "Reason: " << reason << '\n'
+      << "Backtrace:" << '\n'
       << backtrace << '\n';
```
Comment (krobelus) [inline src/util/invariant.cpp]: It seems that the word "Backtrace" occurs twice in the example, so I think we should remove
https://github.com/NlightNFotis/cbmc/blob/invariant_changes/src/util/invariant.cpp#L82

Comment (NlightNFotis) [reply src/util/invariant.cpp]: Nice, I hadn't seen that backtrace was there. Thanks for it.

**Comment by NlightNFotis** [comment]
@krobelus Solid review, thanks for your time.

**Comment by NlightNFotis** [comment]
This has served its purpose as a primary discussion ground for this work, and has now been superseded by the actual PR against develop --> https://github.com/diffblue/cbmc/pull/2666
